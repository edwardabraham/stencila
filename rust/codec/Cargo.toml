[package]
name = "codec"
description = "Defines a specification and trait for codecs"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
stencila-schema = { path = "../schema" }
utils = { path = "../utils" }
