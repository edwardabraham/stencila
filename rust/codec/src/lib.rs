use std::{fs::create_dir_all, path::Path};

use common::{
    async_trait::async_trait,
    eyre::{bail, Result},
    serde::{Deserialize, Serialize},
    tokio::{
        fs::File,
        io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter},
    },
};
use stencila_schema::Node;
use utils::vec_string;

// Re-export for the convenience of crates that implement `CodecTrait`
pub use common;
pub use stencila_schema;
pub use utils;

/// A specification for codecs
///
/// All codecs, including those implemented in plugins, should provide this
/// specification. Rust implementations return a `Codec` instance from the
/// `spec` function of `CodecTrait`. Plugins provide a JSON or YAML serialization
/// as part of their manifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct Codec {
    /// A list of format names (or aliases) that the codec can decode from
    /// or encode to
    pub formats: Vec<String>,

    /// Whether the codec supports decoding from string content
    pub from_string: bool,

    /// Whether the codec supports decoding from a file system path
    pub from_path: bool,

    /// Whether the codec supports encoding to string content
    pub to_string: bool,

    /// Whether the codec supports encoding to a file system path
    pub to_path: bool,

    /// A list of root node types that the codec can encode / decode
    ///
    /// Most codecs usually only handle one root type e.g. `Article`.
    /// Used to provide a list of formats to the user that support
    /// the current document type.
    pub root_types: Vec<String>,

    /// A list of node types that the codec does not support
    ///
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_types: Vec<String>,

    /// A list of node properties that the codec does not support
    ///
    /// The format for these strings is `<type>.<property>` e.g. `Article.funders`
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_properties: Vec<String>,

    /// The status of the codec e.g. `alpha`, `beta`, `stable`
    pub status: String,
}

impl Default for Codec {
    fn default() -> Self {
        Self {
            formats: vec![],
            from_string: true,
            from_path: true,
            to_string: true,
            to_path: true,
            root_types: vec_string!["Article"],
            unsupported_types: vec![],
            unsupported_properties: vec![],
            status: "alpha".to_string(),
        }
    }
}

/// A trait for codecs
///
/// This trait can be used by Rust implementations of codecs, allowing them to
/// be compiled into the Stencila binaries.
///
/// It defines similar functions to `serde_json` (and other `serde_` crates) for
/// converting nodes to/from strings, files, readers etc.
#[async_trait]
pub trait CodecTrait {
    /// Get the [`Codec`] specification for this implementation
    fn spec() -> Codec;

    /// Decode a document node from a string
    fn from_str(_str: &str, _options: Option<DecodeOptions>) -> Result<Node> {
        bail!("Decoding from string is not implemented for this format")
    }

    /// Decode a document node from a string asynchronously
    async fn from_str_async(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
        Self::from_str(str, options)
    }

    /// Decode a document node from a `BufReader`
    async fn from_buffer<T: AsyncRead>(
        reader: &mut BufReader<T>,
        options: Option<DecodeOptions>,
    ) -> Result<Node>
    where
        T: Send + Sync + Unpin,
    {
        let mut content = String::new();
        reader.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file
    ///
    /// This function reads the file as a string and passes that on to `from_str`
    /// for decoding. If working with binary formats, you should override this function
    /// to read the file as bytes instead.
    async fn from_file(file: &mut File, options: Option<DecodeOptions>) -> Result<Node> {
        let mut content = String::new();
        file.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file system path
    async fn from_path(path: &Path, options: Option<DecodeOptions>) -> Result<Node> {
        let mut file = File::open(path).await?;
        Self::from_file(&mut file, options).await
    }

    /// Encode a document node to a string
    fn to_string(_node: &Node, _options: Option<EncodeOptions>) -> Result<String> {
        bail!("Encoding to string is not implemented for this format")
    }

    /// Encode a document node to a string asynchronously
    async fn to_string_async(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
        Self::to_string(node, options)
    }

    /// Encode a document node to a `BufWriter`
    async fn to_buffer<T: AsyncWrite>(
        node: &Node,
        writer: &mut BufWriter<T>,
        options: Option<EncodeOptions>,
    ) -> Result<()>
    where
        T: Send + Sync + Unpin,
    {
        let content = Self::to_string_async(node, options).await?;
        writer.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file
    async fn to_file(node: &Node, file: &mut File, options: Option<EncodeOptions>) -> Result<()> {
        let content = Self::to_string_async(node, options).await?;
        file.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file system path
    async fn to_path(node: &Node, path: &Path, options: Option<EncodeOptions>) -> Result<()> {
        if let Some(parent) = path.parent() {
            create_dir_all(parent)?
        }
        let mut file = File::create(path).await?;
        Self::to_file(node, &mut file, options).await
    }
}

/// Decoding options
///
/// Decoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to.
#[derive(Clone, Default)]
pub struct DecodeOptions {
    /// The format of the input to be decoded
    ///
    /// Most codecs only decode one format. However, for those that handle multiple
    /// format it may be necessary to specify this option.
    pub format: Option<String>,
}

/// Encoding options
///
/// Encoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to. Indeed, some options do not apply for some formats.
/// For example, a PDF is always `standalone` (so if that option is set to `false`, it will be ignored).
/// Futhermore, some combinations of options are ineffectual e.g. a `theme` when `standalone: false`
#[derive(Clone)]
pub struct EncodeOptions {
    /// Whether to encode in compact form
    ///
    /// Some formats (e.g HTML and JSON) can be encoded in either compact
    /// or "pretty-printed" (e.g. indented) forms.
    pub compact: bool,

    /// Whether to ensure that the encoded document is standalone
    ///
    /// Some formats (e.g. Markdown, DOCX) are always standalone.
    /// Others can be fragments, or standalone documents (e.g HTML).
    pub standalone: bool,

    /// Whether to bundle local media files into the encoded document
    ///
    /// Some formats (e.g. DOCX, PDF) always bundle. For HTML, bundling means
    /// including media as data URIs rather than links to files.
    pub bundle: bool,

    /// Whether to use Stencila Cloud as a remote storage for document nodes
    /// that can not be wholly represented within the format
    ///
    /// Some formats (e.g. Google Docs) are not able to fully represent some types
    /// of nodes (e.g. CodeChunks). In these cases we store the node on Stencila Cloud
    /// with a unique id and key and insert a link to it from within the format.
    pub rpng_content: bool,
    pub rpng_link: bool,

    /// The theme to apply to the encoded document
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub theme: Option<String>,

    /// Whether Javascript for Web Components should be loaded or not
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub components: bool,

    /// The format to encode to
    ///
    /// Most codecs only encode to one format. However, for those that handle multiple
    /// formats it may be necessary to specify this option.
    pub format: Option<String>,
}

impl Default for EncodeOptions {
    fn default() -> Self {
        Self {
            compact: true,
            standalone: false,
            bundle: false,
            rpng_content: false,
            rpng_link: false,
            theme: None,
            components: true,
            format: None,
        }
    }
}
