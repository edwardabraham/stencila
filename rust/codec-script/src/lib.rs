use codec::{
    common::{
        eyre::{bail, Result},
        regex::Regex,
    },
    stencila_schema::{Article, BlockContent, CodeChunk, Node},
    utils::vec_string,
    Codec, CodecTrait, DecodeOptions, EncodeOptions,
};
use codec_md::ToMd;
use common::itertools::Itertools;

// A codec for programming language scripts
pub struct ScriptCodec;

impl CodecTrait for ScriptCodec {
    fn spec() -> Codec {
        Codec {
            status: "beta".to_string(),
            formats: vec_string!["bash", "js", "py", "r", "sh", "zsh"],
            root_types: vec_string!["Article"],
            ..Default::default()
        }
    }

    fn from_str(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
        let options = options.unwrap_or_default();
        let lang = match options.format {
            Some(format) => format.to_lowercase(),
            None => bail!("A format option (the programming language of the script) is required"),
        };

        // Define single line comment regexes for each langage
        let single_line_regex = Regex::new(match lang.as_str() {
            "js" => r"^//\s*(.*)$",
            "bash" | "py" | "r" | "sh" | "zsh" => r"^#\s*(.*)$",
            _ => bail!("Unhandled programming language `{}`", lang),
        })
        .expect("Regex should compile");

        // Define multi-line block comment regexes (begin, mid, end)
        let multi_line_regexes = match lang.as_str() {
            "js" => Some((r"^/\*+\s*(.*)$", r"^\s*\*?\s*(.*)$", r"^\s*(.*?)\*+/$")),
            _ => None,
        }
        .map(|regexes| {
            (
                Regex::new(regexes.0).expect("Regex should compile"),
                Regex::new(regexes.1).expect("Regex should compile"),
                Regex::new(regexes.2).expect("Regex should compile"),
            )
        });

        // Split into lines and classify each as either Markdown or code
        // using the regexes
        let mut blocks = Vec::new();
        let md_blocks = |md: &str| -> Vec<BlockContent> {
            codec_md::decode_fragment(md, Some(lang.to_string()))
        };
        let code_block = |code: &str| -> BlockContent {
            BlockContent::CodeChunk(CodeChunk {
                programming_language: lang.clone(),
                text: code.trim().to_string(),
                ..Default::default()
            })
        };
        let mut in_multiline = false;
        let mut md = String::new();
        let mut code = String::new();
        for line in str.lines() {
            if let Some((start_regex, mid_regex, end_regex)) = &multi_line_regexes {
                if in_multiline {
                    let line_md = if let Some(captures) = end_regex.captures(line) {
                        in_multiline = false;
                        captures[1].to_string()
                    } else if let Some(captures) = mid_regex.captures(line) {
                        captures[1].to_string()
                    } else {
                        line.to_string()
                    };
                    md.push_str(&line_md);
                    md.push('\n');
                    continue;
                } else if let Some(captures) = start_regex.captures(line) {
                    if !code.is_empty() {
                        blocks.push(code_block(&code));
                        code.clear();
                    }

                    in_multiline = true;
                    md.push_str(&captures[1]);
                    md.push('\n');
                    continue;
                } else {
                    in_multiline = false;
                }
            }

            // Either add the line to Markdown or to code and if switching between them then
            // add to blocks and clear the buffer.
            if let Some(captures) = single_line_regex.captures(line) {
                if !code.is_empty() {
                    blocks.push(code_block(&code));
                    code.clear();
                }

                md.push_str(&captures[1]);
                md.push('\n');
            } else {
                if !md.is_empty() {
                    blocks.append(&mut md_blocks(&md));
                    md.clear();
                }

                code.push_str(line);
                code.push('\n');
            }
        }

        // Any remaining code to add?
        if !code.is_empty() {
            blocks.push(code_block(&code));
        }

        // Any remaining Markdown to add?
        if !md.is_empty() {
            blocks.append(&mut md_blocks(&md));
        }

        Ok(Node::Article(Article {
            content: Some(blocks),
            ..Default::default()
        }))
    }

    fn to_string(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
        let options = options.unwrap_or_default();

        let blocks = match node {
            Node::Article(Article { content, .. }) => match content {
                Some(blocks) => blocks,
                None => return Ok(String::new()),
            },
            _ => bail!("Unhandled node type `{}`", node.as_ref()),
        };

        let lang = match options.format {
            Some(format) => format.to_lowercase(),
            None => bail!("A format option (the programming language of the script) is required"),
        };
        let comment_start = match lang.as_str() {
            "bash" | "py" | "r" | "sh" | "zsh" => "# ",
            "js" => "// ",
            _ => bail!("Unhandled programming language `{}`", lang),
        };

        // Iterate over blocks, adding `CodeChunk`s as code, and everything else, as Markdown comments
        let mut script = String::new();
        let mut comment_blocks = Vec::new();
        let blocks_to_comment = |blocks: &Vec<&BlockContent>| -> String {
            blocks
                .iter()
                .map(|block| block.to_md().trim_end().to_string())
                .join("\n\n")
                .lines()
                .map(|line| [comment_start, line].concat())
                .join("\n")
        };
        for block in blocks {
            if let BlockContent::CodeChunk(CodeChunk { text, .. }) = block {
                if !comment_blocks.is_empty() {
                    script.push_str(&blocks_to_comment(&comment_blocks));
                    script.push_str("\n\n");

                    comment_blocks.clear();
                }
                script.push_str(text);

                if text.ends_with('\n') {
                    script.push('\n');
                } else {
                    script.push_str("\n\n");
                }
            } else {
                comment_blocks.push(block)
            }
        }

        if !comment_blocks.is_empty() {
            script.push_str(&blocks_to_comment(&comment_blocks))
        }

        Ok(script.trim_end().to_string() + "\n")
    }
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use super::*;
    use test_snaps::{
        insta::{assert_json_snapshot, assert_snapshot},
        snapshot_fixtures_path_content,
    };

    #[test]
    fn decode_and_encode_articles() {
        snapshot_fixtures_path_content("articles/scripts/*", |path: &Path, content| {
            let format = path
                .extension()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string();

            let article = ScriptCodec::from_str(
                content,
                Some(DecodeOptions {
                    format: Some(format.clone()),
                }),
            )
            .unwrap();
            assert_json_snapshot!(article);

            let script = ScriptCodec::to_string(
                &article,
                Some(EncodeOptions {
                    format: Some(format),
                    ..Default::default()
                }),
            )
            .unwrap();
            assert_snapshot!(script);
        });
    }
}
