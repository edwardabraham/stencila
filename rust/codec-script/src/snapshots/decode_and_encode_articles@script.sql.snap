---
source: rust/codec-script/src/lib.rs
assertion_line: 342
expression: article
input_file: fixtures/articles/scripts/script.sql
---
{
  "type": "Article",
  "content": [
    {
      "type": "Paragraph",
      "content": [
        "First paragraph, followed by some code."
      ]
    },
    {
      "type": "CodeChunk",
      "programmingLanguage": "sql",
      "text": "SELECT 1;"
    },
    {
      "type": "Paragraph",
      "content": [
        "Second paragraph followed by some more code, this time multiline,"
      ]
    },
    {
      "type": "CodeChunk",
      "programmingLanguage": "sql",
      "text": "CREATE TABLE table_1(\n    col_1 INT,\n    col_2 TEXT\n);"
    },
    {
      "type": "Paragraph",
      "content": [
        "A multiline comment with a paragraph."
      ]
    },
    {
      "type": "QuoteBlock",
      "content": [
        {
          "type": "Paragraph",
          "content": [
            "And a quote block"
          ]
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        "And a list,"
      ]
    },
    {
      "type": "List",
      "items": [
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                "One"
              ]
            }
          ]
        },
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                "Two"
              ]
            }
          ]
        },
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                "Three"
              ]
            }
          ]
        }
      ],
      "order": "Ascending"
    },
    {
      "type": "CodeChunk",
      "programmingLanguage": "sql",
      "text": "SELECT * FROM table_1"
    },
    {
      "type": "Paragraph",
      "content": [
        "Currently, unlink for other languages, document parameters are not supported in SQL scripts."
      ]
    }
  ]
}
